### PR TITLE
Introduce TypeSet::Iterator that use ctz to iterate type bits faster

### PR DESCRIPTION
**Pull request by kasperl**

### CODE REVIEW

File: src/compiler/propagation/type_set.h [hunk 1]
```diff
@@ -28,6 +28,47 @@ class BlockTemplate;
...
 class TypeSet {
  public:
+  class Iterator {
+   public:
+    Iterator(TypeSet type, int words_per_type)
+        : cursor_(type.bits_)
+        , end_(&type.bits_[words_per_type - 1]) {
+      ASSERT(!type.is_block());
+      uword bits = *cursor_;
+      int base = -1;  // First type is at bit 1.
```
Comment (floitsch) [inline src/compiler/propagation/type_set.h:38]:
```suggestion
      // The first bit of a type-set is used as a marker and
      // guaranteed to be 0 for non-block typesets.
      // The base is thus -1.
      ASSERT((bits & 1) == 0);
      int base = -1;
```

File: src/compiler/propagation/type_set.h [hunk 2]
```diff
@@ -28,6 +28,47 @@ class BlockTemplate;
...
+        base += WORD_BIT_SIZE;
+      }
+      bits_ = bits;
+      base_ = base;
+      return result;
+    }
+
+   private:
+    uword bits_;
+    int base_;
```
Comment (floitsch) [inline src/compiler/propagation/type_set.h:67]:
```suggestion
    int base_;  // The class-index of the first bit in bits_.
```

File: src/compiler/propagation/type_set.cc [hunk 3]
```diff
@@ -97,6 +99,13 @@ bool TypeSet::add_bool(Program* program) {
   return result;
 }
 
+void TypeSet::add_range(unsigned start, unsigned end) {
+  // TODO(kasper): We can make this much faster.
+  for (unsigned type = start; type < end; type++) {
+    add(type);
+  }
```
Comment (erikcorry) [inline src/compiler/propagation/type_set.cc:106]:
```suggestion
  uword* data = &bits_[start / WORD_BIT_SIZE];
  Utils::mark_bits(data, start % WORD_BIT_SIZE, end - start);
```

Comment (erikcorry) [reply src/compiler/propagation/type_set.cc:106]: See https://github.com/toitlang/toit/pull/1272

File: src/compiler/propagation/type_set.h [hunk 4]
```diff
@@ -28,6 +28,47 @@ class BlockTemplate;
...
+      bits_ = bits;
+      base_ = base;
+    }
+
+    bool has_next() const { return bits_ != 0; }
+
+    unsigned next() {
+      ASSERT(has_next());
+      uword bits = bits_;
+      int index = Utils::ctz(bits);
```
Comment (kasperl) [inline src/compiler/propagation/type_set.h:52]:
```suggestion
      // Compute the index of the next bit by counting 
      // trailing zeros (ctz).
      int index = Utils::ctz(bits);
```

**Pull request merged**